tendermint/src/block/height.rs

```diff
@@ -1,27 +1,62 @@
 use crate::error::{Error, Kind};
 use serde::{de::Error as _, Deserialize, Deserializer, Serialize, Serializer};
+use std::convert::TryInto;
 use std::{
     convert::TryFrom,
     fmt::{self, Debug, Display},
     str::FromStr,
 };
+use tendermint_proto::DomainType;
 
 /// Block height for a particular chain (i.e. number of blocks created since
 /// the chain began)
 ///
 /// A height of 0 represents a chain which has not yet produced a block.
 #[derive(Copy, Clone, Eq, Hash, PartialEq, PartialOrd, Ord)]
-pub struct Height(pub u64);
+pub struct Height(u64);
+
+impl DomainType<i64> for Height {}
+
+impl TryFrom<i64> for Height {
+    type Error = Error;
+
+    fn try_from(value: i64) -> Result<Self, Self::Error> {
+        Ok(Height(value.try_into().map_err(|_| Kind::NegativeHeight)?))
```

Comment:
Yes, it's a kind of Rust inherent magic: Height is defined as `Height(u64)`. Hence the value's `try_from()` method is going to try to convert the number to u64 from the original i64. This fails if the value is negative (if it's positive it fits without overflow) so the resultant error is because the height was negative. The TryInto trait's error is not compatible with the defined Self::Error, so it has to be converted.

This can fail miserably if/when we redefine the Height newtype. We can change it to `u64::try_from(value)` which is more explicit and resolves into the same code.